File: years/2021/d09.py
import os.path


testing = False
day = '09'
if testing:
    f_input = os.path.join('inputs', f'{day}-test.txt')
else:
    f_input = os.path.join('inputs', f'input{day}.txt')

def part_one():
    rows = []
    with open(f_input) as f:
        for line in f:
            rows.append([int(x) for x in line.strip()])

    risk_level = 0

    num_rows = len(rows)
    row_length = len(rows[0])
    for i, row in enumerate(rows):
        for j, h in enumerate(row):
            if h == 9:
                continue
            if j > 0:
                if row[j] >= row[j-1]:
                    continue
            if j < row_length - 1:
                if row[j] >= row[j+1]:
                    continue
            if i > 0:
                if row[j] >= rows[i-1][j]:
                    continue
            if i < num_rows - 1:
                if row[j] >= rows[i+1][j]:
                    continue
            risk_level += 1+h
    print(risk_level)

part_one()


def part_two():
    pass

File: years/2023/d09.rb
# frozen_string_literal: true

require_relative 'base'

class Day < Base
  def part_1
    data = File.read(input).split("\n").map(&:split).map { |x| x.map(&:to_i) }
    data.map do |line|
      line.last + find_next_difference(line)
    end.sum
  end

  def part_2
    data = File.read(input).split("\n").map(&:split).map { |x| x.map(&:to_i) }
    data.map do |line|
      line.first - find_first_difference(line)
    end.sum
  end

  private

  TESTING = false
  DAY = '09'
  
  def input
    TESTING ? File.join('inputs', "input#{DAY}-test.txt") : File.join('inputs', "input#{DAY}.txt")
  end

  def find_next_difference(arr)
    differences = find_differences(arr)
    return 0 if differences.all?(&:zero?)
    differences.last + find_next_difference(differences)
  end

  def find_first_difference(arr)
    differences = find_differences(arr)
    return 0 if differences.all?(&:zero?)
    differences.first - find_first_difference(differences)
  end

  def find_differences(arr)
    arr.each_cons(2).map { |a, b| b - a }
  end
end

Day.new.run
# Day.new.benchmark
